File: 0208-implement-trie-prefix-tree/0208-implement-trie-prefix-tree.java
class Trie {
    class PrefixTree {
        HashMap<Character, PrefixTree> hs = new HashMap<>();
        boolean isEnd = false;
    }
    PrefixTree pf;
    public Trie() {
        pf = new PrefixTree();
    }
    
    public void insert(String word) {
        PrefixTree cur = pf;

        for (char c: word.toCharArray()){
            if (!cur.hs.containsKey(c)) cur.hs.put(c, new PrefixTree());
            cur = cur.hs.get(c); 
        }
        cur.isEnd = true;
    }
    
    public boolean search(String word) {
        PrefixTree cur = pf;

        for (char c: word.toCharArray()){
            if (!cur.hs.containsKey(c)) return false;
            cur = cur.hs.get(c); 
        }
        return cur.isEnd;
    }
    
    public boolean startsWith(String prefix) {
        PrefixTree cur = pf;

        for (char c: prefix.toCharArray()){
            if (!cur.hs.containsKey(c)) return false;
            cur = cur.hs.get(c); 
        }
        return true;
        
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.insert(word);
 * boolean param_2 = obj.search(word);
 * boolean param_3 = obj.startsWith(prefix);
 */

File: 0208-implement-trie-prefix-tree/0208-implement-trie-prefix-tree.cs
public class Trie {
    public class PrefixTree{
        public Dictionary<char, PrefixTree> tree = new Dictionary<char, PrefixTree>();
        public bool ending = false;
    }
    public PrefixTree pt;
    public Trie() {
        pt = new PrefixTree();
    }
    
    public void Insert(string word) {
        PrefixTree cur = pt;
        foreach (char c in word.ToCharArray()){
            if (!cur.tree.ContainsKey(c)) cur.tree[c] = new PrefixTree();
            cur = cur.tree[c];
        }
        cur.ending = true;
    }
    public bool Search(string word) {
        PrefixTree cur = pt;
        foreach (char c in word.ToCharArray()){
            if (!cur.tree.ContainsKey(c)) return false;
            cur = cur.tree[c];
        }
        return cur.ending;
    }
    
    public bool StartsWith(string prefix) {
        PrefixTree cur = pt;
        foreach (char c in prefix.ToCharArray()){
            if (!cur.tree.ContainsKey(c)) return false;
            cur = cur.tree[c];
        }

        return true;
    }
}

/**
 * Your Trie object will be instantiated and called as such:
 * Trie obj = new Trie();
 * obj.Insert(word);
 * bool param_2 = obj.Search(word);
 * bool param_3 = obj.StartsWith(prefix);
 */